pkg/embeddedcluster/upgrade.go

```diff
@@ -0,0 +1,380 @@
+			AirGap:                    current.Spec.AirGap,
+			Artifacts:                 artifacts,
+			Config:                    &newcfg,
+			EndUserK0sConfigOverrides: current.Spec.EndUserK0sConfigOverrides,
+			BinaryName:                current.Spec.BinaryName,
+			LicenseInfo:               &embeddedclusterv1beta1.LicenseInfo{IsDisasterRecoverySupported: license.Spec.IsDisasterRecoverySupported},
+		},
+	}
+
+	log.Printf("Starting cluster upgrade to version %s...", newcfg.Version)
+
+	err = runClusterUpgrade(ctx, k8sClient, newins, registrySettings, license, versionLabel)
+	if err != nil {
+		return fmt.Errorf("run cluster upgrade: %w", err)
+	}
+
+	log.Printf("Cluster upgrade to version %s started successfully", newcfg.Version)
+
+	return nil
+}
+
+// runClusterUpgrade will download the new embedded cluster operator binary and run the upgrade
+// command with the provided installation data. This is needed to get the latest
+// embeddedclusterv1beta1 API version. The upgrade command will first upgrade the embedded cluster
+// operator, wait for the CRD to be up-to-date, and then apply the installation object.
+func runClusterUpgrade(
+	ctx context.Context, k8sClient kubernetes.Interface,
+	in *embeddedclusterv1beta1.Installation, registrySettings registrytypes.RegistrySettings,
+	license kotsv1beta1.License, versionLabel string,
+) error {
+	var bin string
+
+	if in.Spec.AirGap {
+		image := in.Spec.Artifacts.AdditionalArtifacts["operator"]
+		if image == "" {
+			return fmt.Errorf("missing operator image in airgap artifacts")
+		}
+
+		b, err := pullUpgradeBinaryFromRegistry(ctx, k8sClient, registrySettings, image)
+		if err != nil {
+			return fmt.Errorf("pull upgrade binary from registry: %w", err)
+		}
+		bin = b
+	} else {
+		b, err := downloadUpgradeBinary(ctx, license, versionLabel)
+		if err != nil {
+			return fmt.Errorf("download upgrade binary: %w", err)
+		}
+		bin = b
+	}
+	defer os.RemoveAll(bin)
+
+	err := os.Chmod(bin, 0755)
+	if err != nil {
+		return fmt.Errorf("chmod upgrade binary: %w", err)
+	}
+
+	installationData, err := k8syaml.Marshal(in)
+	if err != nil {
+		return fmt.Errorf("marshal installation: %w", err)
+	}
+
+	log.Println("Running upgrade command...")
+
+	ctx, cancel := context.WithTimeout(ctx, clusterUpgradeTimeout)
+	defer cancel()
+
+	args := []string{"upgrade"}
+	if in.Spec.AirGap {
+		// TODO(upgrade): local-artifact-mirror-image should be included in the installation object
+		localArtifactMirrorImage, err := getLocalArtifactMirrorImage(ctx, k8sClient, in, registrySettings)
+		if err != nil {
+			return fmt.Errorf("get local artifact mirror image: %w", err)
+		}
+		args = append(args, "--local-artifact-mirror-image", localArtifactMirrorImage)
+	}
+	args = append(args, "--installation", "-")
+
+	cmd := exec.CommandContext(ctx, bin, args...)
+	cmd.Stdin = strings.NewReader(string(installationData))
+	pr, pw := io.Pipe()
+	defer pw.Close()
+	cmd.Stdout = pw
```

Comment:
why not just 
```suggestion
	cmd.Stdout = os.Stdout
```

Reply:
this is equivalent. why would i change it?

Reply:
i meant for simplicity without having the below goroutine entirely.

Reply:
something like this https://github.com/replicatedhq/kots/blob/kots-upgrader/pkg/upgradeservice/process.go#L95-L96

Reply:
I guess your way has the benefit of distinguishing it's the binary's output